doc/man3/RSA_get0_key.pod

```diff
@@ -74,6 +97,11 @@ RSA object.
 RSA_get0_engine() returns a handle to the ENGINE that has been set for
 this RSA object, or NULL if no such ENGINE has been set.
 
+RSA_set_version() sets the version of an RSA object based on B<ver>
+parameter: B<RSA_ASN1_VERSION_MULTI> for multi-prime RSA, B<RSA_ASN1_VERSION_DEFAULT>
+for normal two-prime RSA, as defined in RFC 8017. RSA_get_version() returns the version
+of RSA object B<r>.
```

Comment:
Rather than making this yet another piece of state for the caller to manage (larger API surface, more complexity, more error-prone), why isn't this implicit? The ideal state would be if `rsa->version` didn't exist at all and were computed at serialization-time. That may be tricky with crypto/asn1's table-based design, but you certainly could just keep the field up-to-date as the user calls the relevant setters.

Reply:
> but you certainly could just keep the field up-to-date as the user calls the relevant setters.

This is just why we can't do this: the setters `RSA_set0_key`, `RSA_set0_factors` and `RSA_set0_crt_params` are shared between normal RSA and multi-prime RSA objects. If `rsa->version` is set automatically in `RSA_set0_multi_prime_params` and later someone wants to reuse this same RSA object, there will be no way to set the version back to `normal RSA`. For this reason, we need to provide the setter/getter of version, and if we provide these things, we should just leave it to the callers to set what kind of an RSA object is.  As discussed [here](https://github.com/openssl/openssl/pull/4241#discussion_r141868711).

Reply:
Why can't you just set it back when you call whatever setter clears the extra primes `RSA_set0_multi_prime_params`? Say that passing in zero or something clears the list. Controlling things indirectly with `rsa->version` makes for a rather confusing API. That shouldn't be state at all (i.e. ideally you should delete the field completely; computing it automatically is merely an approximation of that), just a serialization detail.

Note the cached `BN_MONT_CTX` logic means that reusing an `RSA` object like this is pretty questionable anyway. Better to make a new one. (Ideally OpenSSL 1.1.0 would have ditched `RSA_new`, skipped setters, and added functions like `RSA_new_public` and `RSA_new_private`, making `RSA` objects immutable once created. This would have avoided some internal complexity. But so it goes. Maybe for 1.2.0.)

Reply:
Okay, if reuse of an RSA object seems not feasible and should not be encouraged, I guess it's fine to set this version automatically wile calling `RSA_set0_multi_prime_params` and don't need to consider how to set it back...

Reply:
But for `rsa->version`, it's at least useful to check if the corresponding RSA object is normal or multi-prime'd after loading such an object from a PEM file for instance.

Reply:
I'm with David on this one. RSA_set_version is not exactly something to be trusted for application code to handle.

On related side note, I even find the notion of RSA_set0_multi_prime_params setting *additional* primes a bit counter-intuitive. I mean why can't it set *all* the primes instead of excess ones? Was it considered/discussed?

Reply:
Hmmm, I guess `RSA_set_version` could be not exported and removed...

For the side note, one reason is that the 'additional' primes (and related coefficients, exponents...) are treated as a group and there are already APIs to set `p, q, dmp1, dmq1...`, so I choose to continue to treat the additional primes when designing the 'multi-prime' APIs. This is the same for why there is `RSA_get_multi_prime_extra_count` instead of returning the total prime number...

Reply:
As far as I know the idea behind RSA_set0_*() was that you can't set the object with half filled in parameters. You can't set p and keep q NULL. So for multi prime I would expect all the primes to be set at the same time.